chain/chain/src/doomslug.rs

```diff
@@ -204,47 +181,33 @@ impl DoomslugApprovalsTracker {
             Some(approval) => approval,
         };
 
-        self.approved_stake -= self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
-
-        if approval.is_endorsement {
-            self.endorsed_stake -=
-                self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
-        }
+        let stakes = self.account_id_to_stakes.get(&approval.account_id).map_or((0, 0), |x| *x);
+        self.approved_stake_this_epoch -= stakes.0;
+        self.approved_stake_next_epoch -= stakes.1;
     }
 
-    /// Returns whether the block is ready to be produced without waiting for any extra time,
-    /// has crossed the doomslug threshold (and thus can be produced given enough time passed),
-    /// or is not ready to be produced at all.
-    /// This method only returns `ReadyToProduce` if the block has 2/3 approvals and 1/2 endorsements.
-    /// The production readiness due to enough time passing since crossing the threshold should be
-    /// handled by the caller.
+    /// Returns whether the block has enough approvals, and if yes, since what moment it does.
     ///
     /// # Arguments
     /// * now - the current timestamp
     ///
     /// # Returns
-    /// `None` if the block doesn't have enough approvals yet to cross the doomslug threshold
-    /// `PassedThreshold` if the block has enough approvals to pass the threshold, but can't be
-    ///     produced bypassing the timeout
-    /// `ReadyToProduce` if the block can be produced bypassing the timeout (i.e. has 2/3+ of
-    ///     approvals and 1/2+ of endorsements)
+    /// `NotReady` if the block doesn't have enough approvals yet to cross the threshold
+    /// `ReadySince` if the block has enough approvals to pass the threshold, and since when it
+    ///     does
     fn get_block_production_readiness(&mut self, now: Instant) -> DoomslugBlockProductionReadiness {
-        if self.approved_stake > self.total_stake * 2 / 3
-            && self.endorsed_stake > self.total_stake / 2
-        {
-            if self.time_passed_threshold == None {
-                self.time_passed_threshold = Some(now);
-            }
-            DoomslugBlockProductionReadiness::ReadyToProduce(self.time_passed_threshold.unwrap())
-        } else if self.approved_stake > self.total_stake / 2
+        if ((self.approved_stake_this_epoch > self.total_stake_this_epoch * 2 / 3
+            || self.total_stake_this_epoch == 0)
```

Comment:
what does `self.total_stake_this_epoch == 0` represent?

Reply:
It can't happen, I've done it for consistency. Will remove